src/audio/detect_test.c

```diff
@@ -0,0 +1,482 @@
+	memcpy(keyword, ipc_keyword, sizeof(struct sof_ipc_comp_process));
+
+	kd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*kd));
+
+	if (!kd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, kd);
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void test_keyword_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_keyword("test_keyword_free()");
+
+	detect_test_free_parameters(cd);
+	free_mem_load(cd);
+	rfree(cd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int test_keyword_params(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct sof_ipc_comp_config *config = COMP_GET_CONFIG(dev);
+
+	dev->params.frame_fmt = config->frame_fmt;
+
+	/* detect comp uses mono */
+	dev->params.channels = 1;
+
+	dev->frame_bytes = comp_frame_bytes(dev);
+
+	/* calculate period size based on config */
+	cd->period_bytes = dev->frames * dev->frame_bytes;
+
+	return 0;
+}
+
+static int test_keyword_ctrl_set_data(struct comp_dev *dev,
+				      struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct sof_detect_test_config *cfg;
+	size_t bs;
+	int ret = 0;
+
+	/* Check version from ABI header */
+	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, cdata->data->abi)) {
+		trace_keyword_error("test_keyword_cmd_set_data() error: "
+				    "invalid version");
+		return -EINVAL;
+	}
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_ENUM:
+		trace_keyword("test_keyword_cmd_set_data(), SOF_CTRL_CMD_ENUM");
+		break;
+	case SOF_CTRL_CMD_BINARY:
+		trace_keyword("test_keyword_cmd_set_data(), "
+			      "SOF_CTRL_CMD_BINARY");
+
+		if (dev->state != COMP_STATE_READY) {
+			/* It is a valid request but currently this is not
+			 * supported during playback/capture. The driver will
+			 * re-send data in next resume when idle and the new
+			 * configuration will be used when playback/capture
+			 * starts.
+			 */
+			trace_keyword_error("test_keyword_cmd_set_data() "
+					    "error: driver is busy");
+			return -EBUSY;
+		}
+		/* Check and free old config */
+		detect_test_free_parameters(cd);
+
+		/* Copy new config, find size from header */
+		cfg = (struct sof_detect_test_config *)cdata->data->data;
+		bs = cfg->size;
+
+		trace_keyword("test_keyword_cmd_set_data(), blob size = %u",
+			      bs);
+
+		if (bs > SOF_DETECT_TEST_MAX_CFG_SIZE || bs == 0) {
+			trace_keyword_error("test_keyword_cmd_set_data() "
+					    "error: invalid blob size");
+			return -EINVAL;
+		}
+
+		/* Allocate and make a copy of the blob
+		 * and setup the configuration
+		 */
+		cd->config = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, bs);
+		if (!cd->config) {
+			trace_keyword_error("test_keyword_cmd_set_data() "
+					    "error: alloc failed");
+			return -EINVAL;
+		}
+
+		/* Just copy the configuration.
+		 * The component will be initialized in prepare().
+		 */
+		memcpy(cd->config, cdata->data->data, bs);
+		break;
+	default:
```

Comment:
@slawblauciak doesnt the detect also need to support the switch command? this will indicate whether detection is enabled or not

Reply:
Ah, so that's what it was for... Why did it need 2 values?

Reply:
@slawblauciak it actually doesnt need 2 values, just one as detect comp will operate on mono.  My PoC code was just a PoC, so I was ignoring the value for the second channel anyway

Reply:
Apologies... Fixed!